src/rust/ide/lib/view/src/graph_editor/component/edge.rs

```diff
@@ -251,23 +672,45 @@ pub mod back {
 
 macro_rules! define_components {
     ($name:ident {
-        $($field:ident : $field_type:ty),* $(,)?
+        $($field:ident : ($field_type:ty,  $field_shape_type:expr)),* $(,)?
     }) => {
         #[derive(Debug,Clone,CloneRef)]
         #[allow(missing_docs)]
         pub struct $name {
-            pub logger         : Logger,
-            pub display_object : display::object::Instance,
+            pub logger            : Logger,
+            pub display_object    : display::object::Instance,
+            pub shape_view_events : Rc<Vec<ShapeViewEvents>>,
+            shape_type_map        : Rc<HashMap<display::object::Id,ShapeRole>>,
             $(pub $field : component::ShapeView<$field_type>),*
         }
 
         impl $name {
             /// Constructor.
             pub fn new(logger:Logger, scene:&Scene) -> Self {
                 let display_object = display::object::Instance::new(&logger);
-                $(let $field = component::ShapeView::new(Logger::sub(&logger,stringify!($field)),scene);)*
-                $(display_object.add_child(&$field);)*
-                Self {logger,display_object,$($field),*}
+                $(let $field = component::ShapeView::new(
+                    Logger::sub(&logger,stringify!($field)),scene);)*
+                    $(display_object.add_child(&$field);)*
+                let mut shape_view_events:Vec<ShapeViewEvents> = Vec::default();
```

Comment:
a very very minor thing. Just ocnidering our alignment rules, it's always better to write such things as `= Vec::<ShapeViewEvents>::default()` - then left side of `=` is shorter. Doesnt matter here but I thought I will mention it.